File: sdk/javascript/src/alias.js
/**
 * Alias registry — maps human-readable names to block hashes.
 * Aliases let non-technical users reference blocks by name instead of hash.
 *
 * Usage:
 *   const reg = fb.registry()
 *   const farm = reg.create('actor.producer', { name: 'Green Acres' }, {}, { alias: 'farm' })
 *   const wheat = reg.create('substance.ingredient', { name: 'Wheat' }, { source: '@farm' })
 *   // '@farm' is resolved to farm.hash before block creation
 */

const { create, update } = require('./block')

class Registry {
  constructor() {
    this._aliases = new Map()
  }

  /** Register an alias for a hash. */
  set(alias, hash) {
    this._aliases.set(alias, hash)
    return this
  }

  /** Resolve an alias to a hash. Returns the alias itself if not found (pass-through for raw hashes). */
  resolve(aliasOrHash) {
    if (typeof aliasOrHash === 'string' && aliasOrHash.startsWith('@')) {
      const name = aliasOrHash.slice(1)
      const hash = this._aliases.get(name)
      if (!hash) throw new Error(`FoodBlock: unresolved alias "@${name}"`)
      return hash
    }
    return aliasOrHash
  }

  /** Resolve all @aliases in a refs object. */
  resolveRefs(refs) {
    const resolved = {}
    for (const [key, value] of Object.entries(refs)) {
      if (Array.isArray(value)) {
        resolved[key] = value.map(v => this.resolve(v))
      } else {
        resolved[key] = this.resolve(value)
      }
    }
    return resolved
  }

  /** Create a block, resolving any @aliases in refs. Optionally register an alias for the new block. */
  create(type, state = {}, refs = {}, opts = {}) {
    const resolvedRefs = this.resolveRefs(refs)
    const block = create(type, state, resolvedRefs)
    if (opts.alias) {
      this._aliases.set(opts.alias, block.hash)
    }
    return block
  }

  /** Create an update block, resolving @aliases. */
  update(previousHash, type, state = {}, refs = {}, opts = {}) {
    const resolvedPrev = this.resolve(previousHash)
    const resolvedRefs = this.resolveRefs(refs)
    const block = update(resolvedPrev, type, state, resolvedRefs)
    if (opts.alias) {
      this._aliases.set(opts.alias, block.hash)
    }
    return block
  }

  /** Get all registered aliases. */
  get aliases() {
    return Object.fromEntries(this._aliases)
  }

  /** Check if an alias exists. */
  has(alias) {
    return this._aliases.has(alias)
  }

  /** Get the number of registered aliases. */
  get size() {
    return this._aliases.size
  }
}

function registry() {
  return new Registry()
}

module.exports = { registry, Registry }

File: sdk/go/alias.go
package foodblock

import "errors"

// Registry maps human-readable names to block hashes.
type Registry struct {
	aliases map[string]string
}

// NewRegistry creates a new alias registry.
func NewRegistry() *Registry {
	return &Registry{aliases: make(map[string]string)}
}

// Set registers an alias for a hash.
func (r *Registry) Set(alias, hash string) *Registry {
	r.aliases[alias] = hash
	return r
}

// Resolve resolves an alias (prefixed with @) to a hash. Pass-through for raw hashes.
func (r *Registry) Resolve(aliasOrHash string) (string, error) {
	if len(aliasOrHash) > 0 && aliasOrHash[0] == '@' {
		name := aliasOrHash[1:]
		hash, ok := r.aliases[name]
		if !ok {
			return "", errors.New("FoodBlock: unresolved alias \"@" + name + "\"")
		}
		return hash, nil
	}
	return aliasOrHash, nil
}

// ResolveRefs resolves all @aliases in a refs map.
func (r *Registry) ResolveRefs(refs map[string]interface{}) (map[string]interface{}, error) {
	resolved := make(map[string]interface{})
	for key, value := range refs {
		switch v := value.(type) {
		case string:
			h, err := r.Resolve(v)
			if err != nil {
				return nil, err
			}
			resolved[key] = h
		case []interface{}:
			arr := make([]interface{}, len(v))
			for i, item := range v {
				if s, ok := item.(string); ok {
					h, err := r.Resolve(s)
					if err != nil {
						return nil, err
					}
					arr[i] = h
				} else {
					arr[i] = item
				}
			}
			resolved[key] = arr
		default:
			resolved[key] = value
		}
	}
	return resolved, nil
}

// Create creates a block resolving @aliases in refs. Optionally registers an alias for the new block.
func (r *Registry) Create(typ string, state, refs map[string]interface{}, alias string) (Block, error) {
	resolvedRefs, err := r.ResolveRefs(refs)
	if err != nil {
		return Block{}, err
	}
	block := Create(typ, state, resolvedRefs)
	if alias != "" {
		r.aliases[alias] = block.Hash
	}
	return block, nil
}

// UpdateBlock creates an update block resolving @aliases.
func (r *Registry) UpdateBlock(previousHash, typ string, state, refs map[string]interface{}, alias string) (Block, error) {
	resolvedPrev, err := r.Resolve(previousHash)
	if err != nil {
		return Block{}, err
	}
	resolvedRefs, err := r.ResolveRefs(refs)
	if err != nil {
		return Block{}, err
	}
	block := Update(resolvedPrev, typ, state, resolvedRefs)
	if alias != "" {
		r.aliases[alias] = block.Hash
	}
	return block, nil
}

// Aliases returns all registered aliases.
func (r *Registry) Aliases() map[string]string {
	result := make(map[string]string)
	for k, v := range r.aliases {
		result[k] = v
	}
	return result
}

// Has checks if an alias exists.
func (r *Registry) Has(alias string) bool {
	_, ok := r.aliases[alias]
	return ok
}

// Size returns the number of registered aliases.
func (r *Registry) Size() int {
	return len(r.aliases)
}
